File: Python/27. 해외선물-실시간 시세.py
import asyncio
import ebest
from common import *
from app_keys import appkey, appsecretkey # app_keys.py 파일에 appkey, appsecretkey 변수를 정의하고 사용하세요

async def main():
    api=ebest.OpenApi()
    api.on_message = on_message
    api.on_realtime = on_realtime
    if not await api.login(appkey, appsecretkey): return print(f'연결실패: {api.last_message}')

    symbol = 'HSIH24' # 항셍선물 2024년 3월물, 또는 다른 선물 종목코드로 변경하세요.
    
    sym8 = f'{symbol:8}' # 8자리로 맞추기

    # [실시간 시세 요청] OVC : 해외선물 현재가체결
    await api.add_realtime('OVC', sym8)
    
    # 10분후 실시간 시세 중지
    print('10분동안 실시간 작동중...');
    await asyncio.sleep(600)
    await api.remove_realtime('OVC', sym8)
    await asyncio.sleep(1)
    
    ... # 다른 작업 수행
    await api.close()
    

def on_message(api:ebest.OpenApi, msg:str): print(f'on_message: {msg}')

def on_realtime(api:ebest.OpenApi, trcode, key, realtimedata):
    if trcode == 'OVC':
        print(f'on_realtime: 해외선물 체결시세: {trcode}, {key}')
        print_table(realtimedata)
        
asyncio.run(main())

# Output:
'''
10분동안 실시간 작동중...
on_message: OVC(3): 정상처리되었습니다
on_realtime: 해외선물 체결시세: OVC, HSIH24
Field Count = 19
+-----------+----------+
|    key    |  value   |
+-----------+----------+
|  ovsmkend | 20240307 |
|   symbol  |  HSIH24  |
|    lSeq   |    2     |
|  chgrate  |  -0.72   |
|  kordate  | 20240307 |
|   trdtm   |  134718  |
|   curpr   | 16318.0  |
|  ovsdate  | 20240307 |
|  mdvolume |          |
|  ydiffpr  |  118.0   |
|    totq   |  82909   |
|    high   | 16552.0  |
| ydiffSign |    5     |
|    low    | 16296.0  |
|  msvolume |          |
|   cgubun  |    -     |
|    trdq   |    1     |
|    open   | 16438.0  |
|   kortm   |  144718  |
+-----------+----------+
on_realtime: 해외선물 체결시세: OVC, HSIH24
Field Count = 19
+-----------+----------+
|    key    |  value   |
+-----------+----------+
|  ovsmkend | 20240307 |
|   symbol  |  HSIH24  |
|    lSeq   |    3     |
|  chgrate  |  -0.72   |
|  kordate  | 20240307 |
|   trdtm   |  134718  |
|   curpr   | 16318.0  |
|  ovsdate  | 20240307 |
|  mdvolume |          |
|  ydiffpr  |  118.0   |
|    totq   |  82910   |
|    high   | 16552.0  |
| ydiffSign |    5     |
|    low    | 16296.0  |
|  msvolume |          |
|   cgubun  |    -     |
|    trdq   |    1     |
|    open   | 16438.0  |
|   kortm   |  144718  |
+-----------+----------+
...
'''

File: CSharp/27. 해외선물-실시간 시세.cs
using System.Text.Json;

namespace CSharp;

internal class _27 : SampleBase
{
    public override async Task ActionImplement()
    {
        api.OnRealtimeEvent += Api_OnRealtimeEvent;

        // 해외선물 종목코드 입력
        var symbol = await GetInputAsync("해외선물 종목코드를 입력하세요 (ex HSIH24):");

        string symbol_8 = (symbol + "        ")[..8]; // 8자리로 맞추기

        // [실시간 시세 요청] OVC : 해외선물 현재가체결
        if (!await api.AddRealtimeRequest("OVC", symbol_8))
        {
            print($"실시간 시세 요청 실패: {api.LastErrorMessage}");
            return;
        }

        print("실시간 작동중... 중지 할려면 아무키를 누르세요");
        var key = await GetReadKeyAsync();

        await api.RemoveRealtimeRequest("OVC", symbol_8);

        api.OnRealtimeEvent -= Api_OnRealtimeEvent;
    }

    private void Api_OnRealtimeEvent(object? sender, eBEST.OpenApi.Events.EBestOnRealtimeEventArgs e)
    {
        // OnRealtimeEvent 이벤트
        print($"{e.TrCode}, {e.Key}");
        if (e.TrCode.Equals("OVC"))
        {
            var outBlockData = e.RealtimeBody.Deserialize<OVCOutBlock>(_jsonOptions);
            if (outBlockData is not null)
            {
                print(outBlockData);
            }
        }
    }
}

// Output:
/*
해외선물 종목코드를 입력하세요 (ex HSIH24):HSIH24
10분동안 실시간 작동중...
OVC, HSIH24
OVCOutBlock, Field Count = 18
| Key       | Value    |
|-----------|----------|
| symbol    | HSIH24   |
| ovsdate   | 20240308 |
| kordate   | 20240308 |
| trdtm     | 181153   |
| kortm     | 191153   |
| curpr     | 16360    |
| ydiffpr   | 15       |
| ydiffSign | 2        |
| open      | 16345    |
| high      | 16471    |
| low       | 16339    |
| chgrate   | 0.09     |
| trdq      | 1        |
| totq      | 3565     |
| cgubun    | -        |
| mdvolume  |          |
| msvolume  |          |
| ovsmkend  | 20240311 |

OVC, HSIH24
OVCOutBlock, Field Count = 18
| Key       | Value    |
|-----------|----------|
| symbol    | HSIH24   |
| ovsdate   | 20240308 |
| kordate   | 20240308 |
| trdtm     | 181153   |
| kortm     | 191153   |
| curpr     | 16360    |
| ydiffpr   | 15       |
| ydiffSign | 2        |
| open      | 16345    |
| high      | 16471    |
| low       | 16339    |
| chgrate   | 0.09     |
| trdq      | 1        |
| totq      | 3566     |
| cgubun    | -        |
| mdvolume  |          |
| msvolume  |          |
| ovsmkend  | 20240311 |
 */